File: ImageShader.cpp
#include "ImageShader.hpp"

RasterizationData ImageShader::vertexShader(const void * vertexBuffer, const int vertexIdx)
{
	ImageShaderVertex vertex = ((ImageShaderVertex*)vertexBuffer)[vertexIdx];
	RasterizationData out;
	out.position = glm::vec4(vertex.position.x, vertex.position.y, 1.0f, 1.0f);
	out.extraData.push_back(glm::vec4(vertex.uv.x, vertex.uv.y, 1.0f, 1.0f));
	return out;
}

glm::vec4 ImageShader::fragmentShader(const RasterizationData & rasterizationData)
{
	glm::vec4 uv = rasterizationData.extraData[0];
	glm::vec2 _uv = glm::vec2(uv.x, uv.y);
	if (texture)
	{
		glm::vec4 color = texture->sample(_uv);
		return color;
	}
	else
	{
		return glm::vec4(0.0f);
	}
}

File: Rust/example/src/image_shader.rs
use nalgebra::{Vector2, Vector3, Vector4};
use sr::{
    shader::{EShaderAttribute, RasterizationData, Shader, ShaderVertexData},
    texture::Texture,
};
use std::sync::Arc;

pub struct ImageShader {
    pub texture: Arc<Texture>,
}

#[derive(Debug, Clone, Copy)]
pub struct ImageShaderVertex {
    pub position: Vector2<f32>,
    pub uv: Vector2<f32>,
}

impl ShaderVertexData for ImageShaderVertex {}

impl Shader<ImageShaderVertex> for ImageShader {
    fn vertex(&self, vertex_data: &ImageShaderVertex) -> RasterizationData {
        let out = RasterizationData {
            position: Vector4::new(vertex_data.position.x, vertex_data.position.y, 0.0, 1.0),
            attributes: vec![EShaderAttribute::Vec2(vertex_data.uv)],
        };
        out
    }

    fn fragment(&self, rasterization_data: &RasterizationData) -> Vector4<f32> {
        match &rasterization_data.attributes[0] {
            EShaderAttribute::Vec2(uv) => {
                let uv: Vector3<f32> = Vector3::new(uv.x, uv.y, 0.0);
                let color = self.texture.sample(&uv);
                color
            }
            _ => Vector4::from_element(0.0),
        }
    }
}
